### PR TITLE
Fix bug in trimming. Discuss pipelining.

### PR DESCRIPTION
**Pull request by edsko**

### CODE REVIEW

File: ouroboros-consensus/docs/ChainSync.md [hunk 1]
```diff
@@ -33,9 +37,40 @@ the upstream node.
...
+away could potentially be very long, so we have to make a case that throwing
+it away is okay.
+
+The scenario is something like this:
+
+```
+                                k
+                           /~~~~~~~~~~\                                 
+
+..................\........------------ our fragment
```
Comment (mrBliss) [inline ouroboros-consensus/docs/ChainSync.md]: We should clarify that the dashes on this line are the headers that have been added when our chain changed.

File: ouroboros-consensus/docs/ChainSync.md [hunk 2]
```diff
@@ -8,9 +8,10 @@ the most _recent_ intersection (could be genesis).
 0.  We will never roll back more than `k` blocks.
 1.  This means we are not interested in upstream nodes whose chains fork from
     our chain more than `k` blocks ago.
-    (If an upstream node does not satisfy this condition, we will disconnect;
-    it will be the responsibility of the network layer to decide whether or not
-    to try reconnecting to such nodes later.)
+    - If an upstream node does not satisfy this condition, we will disconnect
```
Comment (mrBliss) [inline ouroboros-consensus/docs/ChainSync.md]: End this sentence with a period, for consistency with the other two points :grin:

File: ouroboros-consensus/docs/ChainSync.md [hunk 3]
```diff
@@ -64,10 +99,23 @@ We get notified through the chain sync protocol about updates to their chain:
...
+within `k` from our tip; since our fragment is `k` long, it means the
+intersection will be somewhere along our fragment.
+
+This means that we should allow the remote node to roll back from its current
+intersection point to our chain to any point on our fragment, but we do not need
+to allow it to roll back more than that. This means we can always trim their
+fragment so that it is anchored at our anchor point.
+
+- In the common case that both nodes are following the same chain and both up
+  to date, this would leave fragments identical.
```
Comment (mrBliss) [inline ouroboros-consensus/docs/ChainSync.md]: the/both fragments

File: ouroboros-consensus/docs/ChainSync.md [hunk 4]
```diff
@@ -64,10 +99,23 @@ We get notified through the chain sync protocol about updates to their chain:
...
+- In the extreme case that they are `k` behind, it would leave their fragment
+  empty.
+- More generally, this may leave their fragment shorter than `k`, thus not
+  allowing them their full rollback. However, this is okay: if their fragment
+  _is_ shorter than `k` due to trimming to the intersection point, and they _do_
```
Comment (mrBliss) [inline ouroboros-consensus/docs/ChainSync.md]: intersection point -> anchor point?

Comment (edsko) [reply ouroboros-consensus/docs/ChainSync.md]: ah, yes, well spotted, that was a remnant from our previous attempt.

File: ouroboros-consensus/docs/ChainSync.md (hunk 1 above)
Comment (edsko) [reply ouroboros-consensus/docs/ChainSync.md]: Discussed this on slack, we'll leave it as it.

**Comment by edsko** [comment]
Force-pushed the changes.

**Pull request merged**